src/passes/OptimizeInstructions.cpp

```diff
@@ -974,25 +955,50 @@ struct OptimizeInstructions
   Expression* optimizeBoolean(Expression* boolean) {
     // TODO use a general getFallthroughs
     if (auto* unary = boolean->dynCast<Unary>()) {
-      if (unary && unary->op == EqZInt32) {
-        auto* unary2 = unary->value->dynCast<Unary>();
-        if (unary2 && unary2->op == EqZInt32) {
-          // double eqz
-          return unary2->value;
+      if (unary) {
+        if (unary->op == EqZInt32 || unary->op == EqZInt64) {
+          if (unary->op == EqZInt32) {
+            auto* unary2 = unary->value->dynCast<Unary>();
+            if (unary2 && unary2->op == EqZInt32) {
+              // double eqz
+              return unary2->value;
+            }
+          }
+          if (auto* c = unary->value->dynCast<Const>()) {
+            // [i32 | i64] !C => eval C ? 0 : 1
+            c->value = Literal(int32_t(c->value.getInteger() ? 0 : 1));
+            c->type = Type::i32;
+            return unary->value;
+          }
         }
       }
     } else if (auto* binary = boolean->dynCast<Binary>()) {
-      if (binary->op == OrInt32) {
+      if (binary->op == SubInt32) {
+        if (auto* num = binary->left->dynCast<Const>()) {
+          if (num->value.geti32() == 0) {
+            // 0 - x   ==>   x
```

Comment:
this optimization looks unrelated to the main topic here, of optimizing a select with constant arms? A few other things as well. It would be good to reduce this as much as possible to a small change for easier review.

Reply:
it apply `bool(-x) ==> bool(x)` for `optimizeBoolean` which implicitly relate to `select` and `if` branches. But I agree it better use for separate PR